localization_manager/src/LocalizationTransitionTable.cpp

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (C) 2019-2020 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#include <ros/console.h>
+#include "localization_manager/LocalizationTransitionTable.h"
+
+namespace localizer
+{
+LocalizationTransitionTable::LocalizationTransitionTable(LocalizerMode mode) : mode_(mode)
+{
+}
+LocalizationState LocalizationTransitionTable::getState()
+{
+  return state_;
+}
+
+void LocalizationTransitionTable::logDebugSignal(LocalizationSignal signal)
+{
+  ROS_DEBUG_STREAM("LocalizationTransitionTable received unsupported signal of " << signal << " while in state "
+                                                                                 << state_);
+}
+
+void LocalizationTransitionTable::setAndLogState(LocalizationState new_state, LocalizationSignal source_signal)
+{
+  if (new_state == state_)
+  {
+    return;  // State was unchanged no need to log or trigger callbacks
+  }
+  ROS_INFO_STREAM("LocalizationTransitionTable changed localization state from "
+                  << state_ << " to " << new_state << " because of signal " << source_signal << " while in mode "
+                  << mode_);
+  LocalizationState prev_state = state_;
+  state_ = new_state;
+  if (transition_callback_)
+  {
+    transition_callback_(prev_state, state_, source_signal);
+  }
+}
+
+void LocalizationTransitionTable::signalWhenUNINITIALIZED(LocalizationSignal signal)
+{
+  switch (signal)
+  {
+    case LocalizationSignal::INITIAL_POSE:
+      if (mode_ == LocalizerMode::GNSS)
+      {
+        setAndLogState(LocalizationState::DEGRADED_NO_LIDAR_FIX, signal);
+      }
+      else
+      {
+        setAndLogState(LocalizationState::INITIALIZING, signal);
+      }
+      break;
+    default:
+      logDebugSignal(signal);
+      break;
+  }
+}
+
+void LocalizationTransitionTable::signalWhenINITIALIZING(LocalizationSignal signal)
+{
+  switch (signal)
+  {
+    // How to handle the combined conditions?
+    case LocalizationSignal::GOOD_NDT_FREQ_AND_FITNESS_SCORE:
+      setAndLogState(LocalizationState::OPERATIONAL, signal);
+      break;
+    case LocalizationSignal::POOR_NDT_FREQ_OR_FITNESS_SCORE:
+      setAndLogState(LocalizationState::DEGRADED, signal);
+      break;
+    case LocalizationSignal::LIDAR_SENSOR_FAILURE:
+      setAndLogState(LocalizationState::DEGRADED_NO_LIDAR_FIX, signal);
```

Comment:
On the state diagram, it seems like we need to check mode here?

Reply:
Good catch. A check is needed, but the diagram showed the wrong check. I have updated the diagram and code.